[Nidhi...Patil] iP

patil-np opened the pull request:
# Julie
> _Too many tasks to handle and nowhere to track them? Julie's got your back!_

Julie is a text-based chatbot who organises your responsibilities and lets you track your progress! She is, 

- Text-based
- Beginner friendly 
- Fast and efficient!

## Here's how you can get started
Simply, 

1. Download her from [here](https://github.com/patil-np/ip/releases/tag/A-Jar) 
2. Run the JAR file
3. Start tracking your tasks!

## Current features
Using Julie, you can: 

- [ ] Add miscellaneous todos 
- [ ] Add and manage deadlines 
- [ ] Add and manage events 
## 
### If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```
public static void main(String[] args) {
        new Julie().run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

patil-np committed:
Rename Duke, greet and exit

patil-np committed:
Echo input, exit at bye

patil-np committed:
Echo input, exit at bye

patil-np committed:
Merge pull request #1 from patil-np/detached

Level 1 fixed

patil-np committed:
Create task list

patil-np committed:
Mark and unmark

patil-np committed:
Types of tasks

patil-np committed:
Refactor main method for improved readability

Main method is bulky, which affects the clarity of code.

Let's extract task-specific logic and create separate methods, namely:
* addToDo
* addDeadline
* addEvent
for the main method to focus solely on control flow.

patil-np committed:
Add exception handling for todo, deadline and event

patil-np committed:
Add exception handling for marking and unmarking

patil-np committed:
Add exception handling for unknown phrases

patil-np committed:
Add handling for task deletion

patil-np committed:
Save changes in tasks to hard disk

patil-np committed:
Add changes made to branch-Level-7

K1mcheee commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,18 @@
+public class Deadline extends Task {
+    protected String by;
+    private final String MARKER = "[D]";
```
Usually all uppercase will be used for static final fields, so perhaps it could be `marker` instead, same for the other classes

or another way is to possibly make this a static field, since it will be same for all instances of Deadline

K1mcheee commented on src/main/java/Julie.java:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    public static void addDeadline(String input) throws WrongFormatException{
+        try{
```
Just in case you missed this, its missing a space (as in `try {` ), same for `deleteTask` below

K1mcheee commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,77 @@
+            String description = parts[2];
+
+            Task task;
+            switch (type) {
+                case "T":
+                    task = new ToDo(description);
+                    break;
+                case "D":
+                    task = new Deadline(description, parts[3]);
+                    break;
+                case "E":
+                    task = new Event(description, parts[3], parts[4]);
+                    break;
+                default:
+                    return null;
```
Perhaps you could (if you're using an IDE) configure it to change the `case` blocks to not have auto indentation so it follows the coding standards (can be accessed through settings)

K1mcheee commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public String toString() {
+        return this.MARKER + " " + super.toString() + "(from: " + this.from + " to: " + this.by + ")";
```
(just sharing other alternatives for string printing) if you find adding a lot of "+" is time consuming, I like to use `String.format()`, something like:
```java
String.format("%s %s (from: %s to: %s)", this.MARKER, super.toString(), this.from, this.by)
```

K1mcheee reviewed (commented):
I like how you abstracted out the different tasks and formatted your code! Just a few suggestions on naming and spacing

patil-np committed:
Parse date/time and handle all cases

patil-np committed:
Add date/time handling

qilstiano commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,30 @@
+
+public class Deadline extends Task {
+    protected LocalDateTime by;
+    private static final String MARKER = "[D]";
```
I like that you remembered to name constants in all caps. Nice! 👍

qilstiano commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,28 @@
+import java.io.Serializable;
+public abstract class Task implements Serializable {
+    protected String description;
+    protected boolean isDone;
```
Great that you remembered the convention for naming boolean variables!

qilstiano commented on src/main/java/Julie.java:
```diff
@@ -0,0 +1,260 @@
+        return String.join(" ", java.util.Arrays.copyOfRange(arr, start, stop));
+    }
+
+    public static void ackMessage(Task task) {
```
I think you could potentially change your method name to just acknowledgeMessage instead of ackMessage. It's a bit of a nit but I suppose it could improve readability of the code slightly. Let me know your thoughts! 👍

qilstiano commented on src/main/java/Julie.java:
```diff
@@ -0,0 +1,260 @@
+            } else {
+                if (parts[0].equals("mark")) {
+                    throw new WrongFormatException(
+                            "Sorry, that's the wrong format for marking! Here's the correct format:\n" +
```
I really like that you ensured that you broke to the next line for long strings throughout your project. Keep it up! 💯

qilstiano reviewed (commented):
So far this looks pretty good! I think you could further Julie with some minor code style changes to make the code as readable as it can be to others. Otherwise, it looks awesome, keep up the great work! 👍 💯

patil-np committed:
Create classes to handle UI and TaskList separately

patil-np committed:
Finish further OOP abstraction

patil-np committed:
Complete MoreOOP level

patil-np committed:
Classify into packages

patil-np committed:
Merge branch 'add-gradle-support'

patil-np committed:
Add gradle support

patil-np committed:
Add JUnit tests for tasks

patil-np committed:
Complete A-JUnit

patil-np committed:
Create JAR file

patil-np committed:
Add JavaDoc comments to Julie and UI classes.

patil-np committed:
Add JavaDoc comments to remaining classes in julie

patil-np committed:
Add JavaDoc comments to tasks and exceptions

patil-np committed:
Add JavaDoc comments to commands

patil-np committed:
Add JavaDoc comments to all classes

patil-np committed:
Check for adherence to coding standards

patil-np committed:
Allow users to find tasks from their list

patil-np committed:
Add find functionality

patil-np committed:
Allow users to find tasks from their list

patil-np committed:
Check for style issues

patil-np committed:
Add GUI to the chatbot

patil-np committed:
Add GUI to chatbot

patil-np committed:
Add images to be used in GUI

patil-np committed:
Improve Julie's integration into GUI

patil-np committed:
Do small clean-ups

patil-np committed:
Add assertions

patil-np committed:
All classes: Add more robust refactoring of code

Multiple classes have certain methods and variables that violate the Java Coding Standard.

Refactoring classes to follow standards makes for improved readability.

Let's,
* Extract helper functions from too-long methods
* Edit naming of certain methods and variables

patil-np committed:
Enhance code quality

patil-np committed:
Add assertions

Adding assertions leads to better debugging in the event of any unforeseen errors.

Let's add assertions where required.

patil-np committed:
Fix bugs related to UI and GUI

There are issues with Julie printing out the correct responses in GUI.

Let's,
* Fix this issue by adding better handling for responseBuffer in UI
* Fix critical error in Julie

patil-np committed:
Add ability to recognise and deal with duplicates

Duplicate items of the same type can be added to the TaskList.

Let's,
* Allow Julie to detect when a duplicate is added and alert the user
* Ensure the duplicate task isn't added to the list

patil-np committed:
Fix index-handling

patil-np committed:
Detect and handle duplicates

patil-np committed:
Check Coding Standard

patil-np committed:
Merge CodingStandard

patil-np committed:
Add priority to Tasks

patil-np committed:
Add priority to tasks

patil-np committed:
Set up read-me

patil-np committed:
Update readme

patil-np committed:
Update README.md

patil-np committed:
Edit files

patil-np committed:
Merge minor changes

patil-np committed:
Edit files

patil-np committed:
Enable Assertions

patil-np committed:
Update README.md
